convert2rhel/checks.py

```diff
@@ -57,6 +62,88 @@ def perform_pre_ponr_checks():
     ensure_compatibility_of_kmods()
 
 
+def check_convert2rhel_latest():
+    """Make sure that we are running the latest version of convert2rhel"""
+
+    # get lastest avalible version
+
+    repo_content = (
+        "[convert2rhel]\n"
+        "name=Convert2RHEL Repository\n"
+        "baseurl=https://ftp.redhat.com/redhat/convert2rhel/$releasever/os/\n"
+        "gpgcheck=0\n"
+        "enabled=1\n"
+    )
+
+    repo_dir = os.path.join(utils.TMP_DIR, "convert2rhel_repo")
+    utils.mkdir_p(repo_dir)
+    # repo_dir = os.path.join("/tmp/","convert2rhel_repo")
+    repo_path = os.path.join(repo_dir, "convert2rhel.repo")
+    store_content_to_file(filename=repo_path, content=repo_content)
+
+    # repoquery --setopt=reposdir=$(pwd)/test_repo --releasever=8 convert2rhel
+    raw_output_convert2rhel_versions, _ = run_subprocess(
+        "repoquery --releasever={releasever} --setopt=reposdir={repodir} convert2rhel".format(
+            releasever=system_info.releasever,
+            # releasever="8",
+            repodir=repo_dir,
+        ),
+        print_output=False,
+    )
+
+    #   - if cannot access url: print warning and continue
+    PKG_NEVR = r"\b(\S+)-(?:([0-9]+):)?(\S+)-(\S+)\b"
+    convert2rhel_versions = re.findall(PKG_NEVR, raw_output_convert2rhel_versions, re.MULTILINE)
+
+    latest_version = ("0", "0.00", "0")
+    #   - get latest version from https://ftp.redhat.com/redhat/convert2rhel/.
+    for package_version in convert2rhel_versions:
+        if package_version[0] != "convert2rhel":
+            continue
+
+        ver_compare = rpm.labelCompare(package_version[1:], latest_version)
+
+    # if current version is not latest:
+    ver_compare = rpm.labelCompare(("0", convert2rhel_version, "0"), ("0", latest_version[1], "0"))
+    # - If envar is detected
+    if "CONVERT2RHEL_UNSUPPORTED_VERSION" in os.environ:
```

Comment:
I think the decision tree here needs to be reworked.  Here's how I would approach the problem:

1. What are the conditions that I need to check?
2. Which conditions match up with which outcomes?
3. What data do I have to form each condition?

let's take those one at a time:

## 1. What conditions do I want to check?
In the jira ticket, Michal and Eric list the following things:

- Check whether there's a newer packaged version than what we currently have available
- Stop the conversion in case of conversions to RHEL 7 and 8, warn when converting to RHEL 6
- If the user has set an environment variable, then continue instead of stopping

## 2. Matching the outcomes with the checks

* If the package version is later than the running version
  * If system version <= 6.x
    * warn
  * else #  7 or 8 (or soon, 9)
    * if environment variable is set
      * warn or info and continue
    * else # env var not set
      * error out with an error message that tells why and about the env var
* else # the running version is the same or later than the package version
  * Continue 

Notice that in this case we have a maximum of three levels of nesting in the if-else blocks.  This matches up with our three conditions.  The maximum nesting happens when we find this: [1st level of nesting] if the packaged version is later than the running version and [2nd level of nesting]  if we're converting something > 6 and [3rd level] whether the env var was set or unset.

# 3. What data do I have for each condition?

So the toplevel condition needs to compare the package version with the running version.  In your code (once it's fixed), you should have package version in the `latest_version` variable and running version in the `convert2rhel_version` variable.  You compare them with `rpm.labelCompare()` to set ver_compare (outside of the loop).  So the toplevel condition will be `if ver_compare == SOMETHING`. (I don't know off the top of my head whether 1 or -1 will mean the running version or package version is later.  You should be able to test that at the python shell, though.)

The second level condition needs to test the system version.  You've found out where to get that, if `system_info.version.major`.

The third level condition tests whether the env var is set.  You've figured out that condition too: `if "CONVERT2RHEL_UNSUPPORTED_VERSION" in os.environ`

## 4. Put it together.

So the decision tree that you have here is a little different than what I came up with in step 2.  What you have right now is:
* if env var:
  *  continue
* else if package version is later
  * if system is <= 6
    * stop conversion
  * else: # system is 7, 8, 9, etc
    * continue
* elif running version is later than the package version:
  * continue with message
* else running version is equal to package version
  * continue with message that is almost the same as the previous one

So there's a few conditions that should move to different nesting levels and a few conditions that are backwards.  For the nesting levels, one thing that is helpful is to look at what you are comparing at each level.  If you are comparing different things (your topmost level compares envvar, ver_compare,  and ver_compare) then ask yourself (1) if you may have made a mistake or (2) if it would be easier to understand if you wrote it a different way.

For conditions being reversed, that happens to everyone from time to time.  You just have to either double check your work or write a simple test case and then see if the code does the opposite of what the test case tells you you wanted to happen.

Reply:
Bravo! I was looking at this part yesterday for almost 1h trying to figure out how to point out a better way of working with this decision tree, but couldn't find any way of writing it down.

Reply:
That makes a lot more since now looking at it that way, in the past few days revisited this code and was trying to see if there was a better way to handle this! Thank you guys I will make changes to the code and push them when i'm done!